File: T-Bank/Algorithms-and-Data-Structures-2025/5sprint-Strings/contest/B-Редактор/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(Q(L + N))
    Memory Complexity O(L + N)
    """
    text: str = sys.stdin.readline().rstrip()
    n: int = int(sys.stdin.readline().rstrip())
    for _ in range(n):
        pattern: str = sys.stdin.readline().rstrip()
        S: str = pattern + '#' + text
        lcp: list[int] = [0] * len(pattern)
        prev_k: int = 0
        freq: int = 0
        positions: list[int] = []
        for idx in range(1, len(S)):
            k: int = prev_k
            while k and S[k] != S[idx]: k = lcp[k - 1]
            if S[k] == S[idx]: k += 1
            if idx < len(pattern): lcp[idx] = k
            prev_k = k 
            if k == len(pattern):
                freq += 1
                positions.append(idx - 2 * len(pattern))
        sys.stdout.write('{} {}\n'.format(freq, ' '.join(str(pos) for pos in positions)))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/5sprint-Strings/contest/B-Редактор/solution.cpp
#include <iostream>
#include <string>
#include <vector>


using namespace std;


void solution() {
    /*
    Time Complexity O(Q(N+L))
    Memory Complexity O(N + L)
    */
    string text;
    cin >> text;
    int n;
    scanf("%d", &n);
    for (int i = 0; i < n; ++i) {
        string pattern;
        cin >> pattern;
        string S = pattern + "#" + text;
        vector<int> lcp (pattern.size());
        int freq = 0;
        vector<int> positions;
        int prevLcp = 0;
        for (int idx = 1; idx < S.size(); ++idx) {
            int k = prevLcp;
            while (k && S[k] != S[idx]) k = lcp[k - 1];
            if (S[k] == S[idx]) ++k;
            prevLcp = k;
            if (idx < pattern.size()) lcp[idx] = k;
            if (k == pattern.size()) {
                ++freq;
                positions.push_back(idx - 2 * pattern.size());
            }
        }
        printf("%d", freq);
        for (int& pos : positions) printf(" %d", pos);
        printf("\n");
    }
}


int main() {
    solution();
}